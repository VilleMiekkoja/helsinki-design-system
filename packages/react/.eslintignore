**/*.d.ts
webpack.config.js
.templates
lib/
