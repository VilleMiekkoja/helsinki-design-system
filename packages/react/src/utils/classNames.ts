import classnames from 'classnames/dedupe';

export default classnames;
