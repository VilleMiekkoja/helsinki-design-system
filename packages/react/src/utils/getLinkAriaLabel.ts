import React from 'react';

export default (
  ariaLabel: string,
  label: React.ReactNode,
  openInNewTabAriaLabel?: string,
  openInExternalDomainAriaLabel?: string,
  external?: boolean,
  openInNewTab?: boolean,
) => {
  let childrenText = ariaLabel || label;
  const newTabText = openInNewTab ? openInNewTabAriaLabel || 'Avautuu uudessa välilehdessä.' : '';
  const externalText = external ? openInExternalDomainAriaLabel || 'Siirtyy toiseen sivustoon.' : '';

  if (childrenText && childrenText.slice(-1) !== '.') {
    childrenText = `${childrenText}.`;
  }

  return [childrenText, newTabText, externalText].filter((text) => text).join(' ');
};
