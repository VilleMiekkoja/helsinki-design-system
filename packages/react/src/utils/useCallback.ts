import { EventHandler, useCallback } from 'react';

export const useCallbackIfDefined = (callback) => {
  const handler = (event) => {
    if (callback) callback(event);
  };

  return useCallback(handler, [callback]);
};

export const useEnterOrSpacePressCallback = (callback) => {
  const handler = (event) => {
    if (event.key === 'Enter' || event.key === ' ') callback(event);
  };

  return useCallbackIfDefined(handler);
};

export const withDefaultPrevented = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.preventDefault();
    return callback.call(callback, event);
  };

export const withPropagationStopping = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.stopPropagation();
    return callback.call(callback, event);
  };
