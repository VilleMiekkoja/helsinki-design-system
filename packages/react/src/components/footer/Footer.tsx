import React from 'react';

// import core base styles
import 'hds-core';
import styles from './Footer.module.scss';
import { Koros, KorosType } from '../koros';
import classNames from '../../utils/classNames';
import { FooterNavigation } from './footerNavigation/FooterNavigation';
import { FooterNavigationGroup } from './footerNavigationGroup/FooterNavigationGroup';
import { FooterNavigationHeading } from './footerNavigationHeading/FooterNavigationHeading';
import { FooterNavigationLink } from './footerNavigationLink/FooterNavigationLink';
import { FooterUtilities } from './footerUtilities/FooterUtilities';
import { FooterUtilityGroup } from './footerUtilityGroup/FooterUtilityGroup';
import { FooterSoMe } from './footerSoMe/FooterSoMe';
import { FooterBase } from './footerBase/FooterBase';
import { FooterCustom } from './footerCustom/FooterCustom';
import { FooterTheme } from './Footer.interface';
import { getComponentFromChildren } from '../../utils/getChildren';
import { useTheme } from '../../hooks/useTheme';

export type FooterProps = React.PropsWithChildren<{
  /**
   * Additional class names to apply to the footer
   */
  className?: string;
  /**
   * Props that will be passed to the native `<footer>` element
   */
  footerProps?: React.ComponentPropsWithoutRef<'footer'>;
  /**
   * Koros type to use in the footer
   */
  korosType?: KorosType;
  /**
   * Defines the footer theme
   */
  theme?: FooterTheme;
  /**
   * The title of the service shown on top of the the footer
   */
  title?: React.ReactNode;
}>;

export const Footer = ({
  children,
  className,
  footerProps,
  korosType = 'basic',
  theme = 'light',
  title,
}: FooterProps) => {
  // custom theme class that is applied to the root element
  const customThemeClass = useTheme<FooterTheme>(styles.footer, theme);
  // filter out navigation from other children so that they can be rendered separately
  const [navigation, childrenWithoutNavigation] = getComponentFromChildren(children, 'FooterNavigation');

  return (
    <footer
      {...footerProps}
      className={classNames(
        styles.footer,
        styles[`koros-${korosType}`],
        typeof theme === 'string' && styles[`theme-${theme}`],
        customThemeClass,
        className,
      )}
    >
      <Koros className={classNames(styles.koros, styles[korosType])} type={korosType} />
      <div className={styles.footerContent}>
        <section className={classNames(styles.navigationContainer, styles.centeredNavigation)}>
          <div className={styles.titleWrapper}>{title && <h2 className={styles.title}>{title}</h2>}</div>
          {navigation}
        </section>
        {childrenWithoutNavigation}
      </div>
    </footer>
  );
};

Footer.Navigation = FooterNavigation;
Footer.NavigationGroup = FooterNavigationGroup;
Footer.NavigationHeading = FooterNavigationHeading;
Footer.NavigationLink = FooterNavigationLink;
Footer.Utilities = FooterUtilities;
Footer.UtilityGroup = FooterUtilityGroup;
Footer.SoMe = FooterSoMe;
Footer.Base = FooterBase;
Footer.Custom = FooterCustom;
