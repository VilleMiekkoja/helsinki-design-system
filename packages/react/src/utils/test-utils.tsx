import React, { PropsWithChildren } from 'react';

import { Navigation } from '../components/navigation';
import { Footer } from '../components/footer';
import { Header } from '../components/header';

export const NavigationWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Navigation menuToggleAriaLabel="menu" skipTo="#content" skipToContentLabel="Skip to content" title="Foo">
    {children}
  </Navigation>
);

export const FooterWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">{children}</Footer>
);

export const FooterNavigationWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Navigation>{children}</Footer.Navigation>
  </Footer>
);

export const FooterNavigationGroupsWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Navigation>
      <Footer.NavigationGroup>{children}</Footer.NavigationGroup>
    </Footer.Navigation>
  </Footer>
);

export const FooterUtilitiesWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Utilities>{children}</Footer.Utilities>
  </Footer>
);

export const HeaderNavigationMenuWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Header>
    <Header.NavigationMenu>{children}</Header.NavigationMenu>
  </Header>
);
