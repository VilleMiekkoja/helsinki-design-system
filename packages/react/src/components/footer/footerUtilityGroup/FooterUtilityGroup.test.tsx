import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Footer } from '../Footer';
import { FooterUtilitiesWrapper } from '../../../utils/test-utils';
import { FooterVariant } from '../Footer.interface';
import { IconFacebook } from '../../../icons';

describe('<Footer.UtilityGroup /> spec', () => {
  it('renders the utility groups', () => {
    const { asFragment } = render(
      <>
        <Footer.UtilityGroup>
          <Footer.GroupHeading
            href="https://google.com"
            onClick={(e) => e.preventDefault()}
            label="Main Page"
            variant={FooterVariant.Utility}
          />
          <Footer.NavigationLink
            href="https://google.com"
            onClick={(e) => e.preventDefault()}
            label="Sub page"
            variant={FooterVariant.Utility}
          />
          <Footer.NavigationLink
            href="https://google.com"
            onClick={(e) => e.preventDefault()}
            label="Sub page"
            variant={FooterVariant.Utility}
          />
        </Footer.UtilityGroup>
        <Footer.UtilityGroup key={6}>
          <Footer.GroupHeading label="Social media" variant={FooterVariant.Utility} />
          <Footer.NavigationLink
            title="Helsingin kaupungin Facebook-tili"
            label="Facebook"
            aria-label="Helsingin kaupungin Facebook-tili"
            icon={<IconFacebook aria-hidden="true" />}
            href="https://facebook.com/helsinginkaupunki/"
          />
          <Footer.NavigationLink
            title="Helsingin kaupungin Facebook-tili"
            label="Facebook"
            aria-label="Helsingin kaupungin Facebook-tili"
            icon={<IconFacebook aria-hidden="true" />}
            href="https://facebook.com/helsinginkaupunki/"
          />
        </Footer.UtilityGroup>
      </>,
      { wrapper: FooterUtilitiesWrapper },
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(
      <>
        <Footer.UtilityGroup ariaLabel="Utility links group">
          <Footer.GroupHeading
            href="https://google.com"
            onClick={(e) => e.preventDefault()}
            label="Main Page"
            variant={FooterVariant.Utility}
          />
          <Footer.NavigationLink
            href="https://google.com"
            onClick={(e) => e.preventDefault()}
            label="Sub page"
            variant={FooterVariant.Utility}
          />
          <Footer.NavigationLink
            href="https://google.com"
            onClick={(e) => e.preventDefault()}
            label="Sub page"
            variant={FooterVariant.Utility}
          />
        </Footer.UtilityGroup>
        <Footer.UtilityGroup key={6} ariaLabel="Social media links">
          <Footer.GroupHeading label="Social media" variant={FooterVariant.Utility} />
          <Footer.NavigationLink
            title="Helsingin kaupungin Facebook-tili"
            label="Facebook"
            aria-label="Helsingin kaupungin Facebook-tili"
            icon={<IconFacebook aria-hidden="true" />}
            href="https://facebook.com/helsinginkaupunki/"
          />
          <Footer.NavigationLink
            title="Helsingin kaupungin Facebook-tili"
            label="Facebook"
            aria-label="Helsingin kaupungin Facebook-tili"
            icon={<IconFacebook aria-hidden="true" />}
            href="https://facebook.com/helsinginkaupunki/"
          />
        </Footer.UtilityGroup>
      </>,
      { wrapper: FooterUtilitiesWrapper },
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
