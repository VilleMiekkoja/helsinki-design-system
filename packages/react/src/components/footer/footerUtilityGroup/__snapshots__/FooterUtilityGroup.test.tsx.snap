// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer.UtilityGroup /> spec renders the utility groups 1`] = `
<DocumentFragment>
  <footer
    class="footer koros-basic theme-light"
  >
    <div
      class="koros basic koros basic"
    >
      <svg
        aria-hidden="true"
        height="85"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            height="85"
            id="koros_basic-1"
            patternUnits="userSpaceOnUse"
            width="106"
            x="0"
            y="0"
          >
            <path
              d="M0,800h20V0c-4.9,0-5,2.6-9.9,2.6S5,0,0,0V800z"
              transform="scale(5.3)"
            />
          </pattern>
        </defs>
        <rect
          fill="url(#koros_basic-1)"
          height="85"
          width="100%"
        />
      </svg>
    </div>
    <div
      class="footerContent"
    >
      <section
        class="navigationContainer centeredNavigation"
      >
        <div
          class="titleWrapper"
        >
          <h2
            class="title"
          >
            Bar
          </h2>
        </div>
      </section>
      <div
        class="utilities"
      >
        <hr
          aria-hidden="true"
          class="divider"
        />
        <div
          class="links widerLinks"
        >
          <div
            class="utilityGroup"
          >
            <div
              class="utilityGroup"
            >
              <a
                class="heading utility"
                href="https://google.com"
              >
                Main Page
              </a>
              <a
                class="item utility"
                href="https://google.com"
              >
                <span>
                  Sub page
                </span>
              </a>
              <a
                class="item utility"
                href="https://google.com"
              >
                <span>
                  Sub page
                </span>
              </a>
            </div>
          </div>
          <div
            class="utilityGroup"
          >
            <div
              class="utilityGroup"
            >
              <span
                class="heading utility"
              >
                Social media
              </span>
              <a
                aria-label="Helsingin kaupungin Facebook-tili"
                class="item"
                href="https://facebook.com/helsinginkaupunki/"
                title="Helsingin kaupungin Facebook-tili"
              >
                <svg
                  aria-hidden="true"
                  class="icon s"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    fill="none"
                    fill-rule="evenodd"
                  >
                    <path
                      d="M0 0h24v24H0z"
                    />
                    <path
                      d="M19 3a2 2 0 012 2v14a2 2 0 01-2 2h-3.356v-7.12h2.435l.35-2.716h-2.785V9.43c0-.788.217-1.322 1.346-1.322h1.439V5.68l-.35-.036-.341-.025c-.38-.026-.873-.048-1.407-.048-2.073 0-3.495 1.266-3.495 3.592v2.005h-2.443v2.716h2.447L12.839 21H5a2 2 0 01-2-2V5a2 2 0 012-2h14z"
                      fill="currentColor"
                    />
                  </g>
                </svg>
                <span>
                  Facebook
                </span>
              </a>
              <a
                aria-label="Helsingin kaupungin Facebook-tili"
                class="item"
                href="https://facebook.com/helsinginkaupunki/"
                title="Helsingin kaupungin Facebook-tili"
              >
                <svg
                  aria-hidden="true"
                  class="icon s"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    fill="none"
                    fill-rule="evenodd"
                  >
                    <path
                      d="M0 0h24v24H0z"
                    />
                    <path
                      d="M19 3a2 2 0 012 2v14a2 2 0 01-2 2h-3.356v-7.12h2.435l.35-2.716h-2.785V9.43c0-.788.217-1.322 1.346-1.322h1.439V5.68l-.35-.036-.341-.025c-.38-.026-.873-.048-1.407-.048-2.073 0-3.495 1.266-3.495 3.592v2.005h-2.443v2.716h2.447L12.839 21H5a2 2 0 01-2-2V5a2 2 0 012-2h14z"
                      fill="currentColor"
                    />
                  </g>
                </svg>
                <span>
                  Facebook
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
